### PR TITLE
JVM Ergonomics Profiles

### PR DESCRIPTION
**Pull request by brunoborges**
Reviewers, please check our [Contributing Guide](../CONTRIBUTING.md) has been followed 
before accepting this PR.

### CODE REVIEW

**Review by karianna — CHANGES_REQUESTED** [review]
I've made a few suggested changes, but overall I think this could be improved by writing in an active voice.  Recommend a GPT pass to make it so :-)

File: ergonomics_profiles/readme.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
+Summary
+-------
+
+Introduce ergonomics profiles and add a `balanced` option, for the existing heuristics, and a `dedicated` option for when the JVM is running on systems with dedicated resources for the one JVM process.
```
[the remark this answers is not in the document]
Comment (brunoborges) [reply ergonomics_profiles/readme.md:4]: I'll keep this `balanced` for now, but will review back later.

File: ergonomics_profiles/readme.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+Goals
+-----
+
+1. Introduce the concept of ergonomics profiles.
+1. Introduce a new flag to select an ergonomic profile.
+1. Define existing heuristics and ergonomics as `balanced`.
+1. Introduce a `dedicated` profile designed for systems with dedicated resources not expected to be shared with other processes.
+1. Automatically select `dedicated` when the JVM believes the system is dedicated to it.
```
[the remark this answers is not in the document]
Comment (brunoborges) [reply ergonomics_profiles/readme.md]: Done: [7fd7e8c](https://github.com/microsoft/openjdk-proposals/pull/12/commits/7fd7e8cc1de80e4557b88af99c327633f72b64c6)

File: ergonomics_profiles/readme.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (brunoborges) [reply ergonomics_profiles/readme.md]: Alright, I changed to `shared` in [7fd7e8c](https://github.com/microsoft/openjdk-proposals/pull/12/commits/7fd7e8cc1de80e4557b88af99c327633f72b64c6)

File: ergonomics_profiles/readme.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+If the JVM beleives it is running in a dedicated environment (e.g., containers) then the `dedicated` profile will be activated.
+
+**_Shared profile_**
+
+The `shared` profile is what the HotSpot JVM does today in terms of default ergonomics.
+
+**_Dedicated profile_**
+
+The `dedicated` profile will contain different heuristics aimed at maximizing resource consumption in the environment with the assumption that the environment is dedicated to the JVM.
```
Comment (karianna) [inline ergonomics_profiles/readme.md:128]:
```suggestion
The `dedicated` profile will contain different heuristics aimed at maximizing resource consumption in the environment, assuming that the environment is dedicated to the JVM.
```

File: ergonomics_profiles/readme.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+This profile will maximize heap size allocation, optimize garbage collector selection with a larger set of options and considerations, vary the active processor counting, set different values for garbage collector threads, more aggressively size native memory expectation, and size internal JVM thread pools accordingly.
```
Comment (karianna) [inline ergonomics_profiles/readme.md:130]:
```suggestion
This profile will maximize heap size allocation, optimize garbage collector selection with a more extensive set of options and considerations, vary the active processor counting, set different values for garbage collector threads, more aggressively size native memory expectation, and size internal JVM thread pools accordingly.
```

File: ergonomics_profiles/readme.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+```java
+var ergonomicsProfile = System.getProperty("java.vm.ergonomics.profile");
+```
+
+The profile selection may also be obtained programmatically through JMX by the inclusion of the following method in `java.management.RuntimeMXBean`:
+
+    String getJvmErgonomicsProfile()
+
+To allow the backporting of this feature to older versions of OpenJDK, the value may also be obtained through the `MBeanServerConnection.getAttribute` method:
```
Comment (karianna) [inline ergonomics_profiles/readme.md:152]:
```suggestion
The value may also be obtained through the `MBeanServerConnection.getAttribute` method to allow the backporting of this feature to older versions of OpenJDK:
```

File: ergonomics_profiles/readme.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+MBeanServerConnection mbs = ...
+
+var oname = new ObjectName(ManagementFactory.RUNTIME_MXBEAN_NAME);
+var ergonomicsProfile = mbs.getAttribute(oname, "JvmErgonomicsProfile");
+```
+
+Alternatives
+------------
+
+At this moment, the only alternative to this proposal is by manualling tuning the JVM. This exercise is very often complex, difficult, and time consuming for Java developers.
```
Comment (karianna) [inline ergonomics_profiles/readme.md:164]:
```suggestion
Currently, the only alternative to this proposal is manually tuning the JVM. This exercise is often complex, challenging, and time-consuming for Java developers.
```

File: ergonomics_profiles/readme.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+Testing
+-------
+
+// What kinds of test development and execution will be required in order
+// to validate this enhancement, beyond the usual mandatory unit tests?
```
Comment (karianna) [inline ergonomics_profiles/readme.md:170]:
```suggestion
// to validate this enhancement beyond the usual mandatory unit tests?
```

File: ergonomics_profiles/readme.md [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+---------------------
+
+None at this time.
+
+Dependencies
+------------
+
+List of related issues, in no particular order, that may have to be addressed before this JEP, or that this JEP may have address as part of its implementation:
+
+1. [JDK-8261242: [Linux] OSContainer::is_containerized() returns true when run outside a container](https://bugs.openjdk.org/browse/JDK-8261242)
```
Comment (karianna) [inline ergonomics_profiles/readme.md:183]:
```suggestion
1. [JDK-8261242: \[Linux\] OSContainer::is_containerized() returns true when run outside a container](https://bugs.openjdk.org/browse/JDK-8261242)
```

File: ergonomics_profiles/readme.md [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+Risks and Assumptions
+---------------------
+
+None at this time.
+
+Dependencies
+------------
+
+List of related issues, in no particular order, that may have to be addressed before this JEP, or that this JEP may have address as part of its implementation:
```
Comment (karianna) [inline ergonomics_profiles/readme.md:181]:
```suggestion
List of related issues, in no particular order, that may have to be addressed before this JEP or that this JEP may have to address as part of its implementation:
```

File: ergonomics_profiles/readme.md [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+Success Metrics
+---------------
+
+No specific success metrics needed.
+
+Motivation
+----------
+
+The default JVM ergonomics were designed to be balanced with shared resources, with the understanding that it would share resources with other processes (e.g., a data store) running on a shared environment, such as a bare metal server or a large virtual machine.
```
Comment (karianna) [inline ergonomics_profiles/readme.md:28]:
```suggestion
The default JVM ergonomics were designed to share resources with other processes (e.g., a data store) running on a shared environment, such as a bare metal server or a large virtual machine.
```

File: ergonomics_profiles/readme.md [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+
+On a study done by an APM vendor (New Relic), it was identified that more than 60% of monitored JVMs in production are running inside environments with resources dedicated to the JVM (i.e., In containers), as opposed to being shared. A large percentage of those JVMs with dedicated resources were running without explicit JVM tuning flags. Therefore the JVM was running with default ergonomics that are traditionally aimed at shared environments.
```
Comment (karianna) [inline ergonomics_profiles/readme.md:30]:
```suggestion
On a study done by an APM vendor (New Relic), it was identified that more than 60% of monitored JVMs in production are running inside environments with resources dedicated to the JVM (i.e., In containers), as opposed to being shared. A large percentage of those JVMs with dedicated resources were running without explicit JVM tuning flags. Therefore the JVM was running with default ergonomics traditionally aimed at shared environments.
```

File: ergonomics_profiles/readme.md [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+
+**_Min heap size_**
+
+| Available memory | Default      |
+|------------------|--------------|
+| 64 MB – 8192 MB  | 8 MB         |
+
+It is possible to observe that these amounts don't adequately map to the intended resource plan of dedicated environments, where the user may have already considered to allocate e.g. 4GB of memory to the JVM, and expect it to use 4GB of memory. The JVM will instead use, in this example, 1GB of memory, and the user will have to manually tune the JVM to use the expected amount of memory.
+
+Further on, it is likely that the JVM does not need to reclaim heap memory as soon as it does in shared environments, as the JVM is the only process running on the system. This means that the JVM may set the initial heap size the same as maximum heap size, while having a fair minimum heap size for other memory pools consumption (e.g. native memory). This shall hint to the garbage collector to delay, or act more lazily, on the action of cleaning and reclaiming heap.
```
Comment (karianna) [inline ergonomics_profiles/readme.md:67]:
```suggestion
Further on, it is likely that the JVM can reclaim heap memory later than it does in shared environments, as the JVM is the only process running on the system. This means that the JVM may set the initial heap size as maximum heap size, while having a suitable minimum heap size for other memory pools consumption (e.g. native memory). This shall signal to the garbage collector to delay, or act more lazily, on the action of cleaning and reclaiming heap.
```

**Review by karianna — COMMENTED** [review]
Lots of suggested edits for ease of reading.

File: ergonomics_profiles/readme.md [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+
+Introduce a new JVM feature for Ergonomics Profiles, with a `shared` profile for the existing heuristics and a `dedicated` option for when the JVM is running on systems with dedicated resources for the one JVM process.
+
+Goals
+-----
+
+1. Introduce the concept of ergonomics profiles.
+1. Introduce a new flag to select an ergonomic profile.
+1. Define existing heuristics and ergonomics as `shared`.
+1. Introduce a `dedicated` profile designed for systems where the JVM is effectively the only process using resources, e.g. In a canonical container deployment scenario.
```
[the remark this answers is not in the document]
Comment (brunoborges) [reply ergonomics_profiles/readme.md:12]: Perhaps a better way to day is "is the primary process"

**Review by karianna — APPROVED** [review]

File: ergonomics_profiles/readme.md [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+
+We propose adding the concept of Ergonomics Profiles while naming the existing default ergonomics as `shared`, and adding a second profile called `dedicated` for when the JVM is to run under environments with resources dedicated to it. New profiles may be added in the future.
+
+**_Selecting a profile_**
+
+To select a profile, we propose a new flag:
+
+    -XX:ErgonomicsProfile=<shared|dedicated>
+
+Users may also select a profile by setting this flag in the environment variable `JAVA_TOOL_OPTIONS`.
```
Comment (DanHeidinga) [inline ergonomics_profiles/readme.md:114]: Is there a reason to call out this env var and not others like `JDK_JAVA_OPTIONS`?

**Review by DanHeidinga — COMMENTED** [review]
Some late comments after reading the draft

**Pull request merged**